modules/dlsobackend/dlsobackend.cc

```diff
@@ -0,0 +1,722 @@
+  if (this->api->abi_version != PDNS_DLSO_ABI_VERSION) {
+    if (this->api->release != nullptr)
+      this->api->release(this->api);
+
+    dlclose(this->dlhandle);
+    throw PDNSException("Failed to initialize dlso, library does not use the same abi version than pdns");
+  }
+
+  // Sanity checks
+  if (this->api->get == nullptr) {
+    if (this->api->release != nullptr)
+      this->api->release(this->api);
+
+    dlclose(this->dlhandle);
+    throw PDNSException("Failed to initialize dlso, lib did not register a mandatory get function");
+  }
+  if (this->api->lookup == nullptr) {
+    if (this->api->release != nullptr)
+      this->api->release(this->api);
+
+    dlclose(this->dlhandle);
+    throw PDNSException("Failed to initialize dlso, lib did not register a mandatory lookup function");
+  }
+}
+
+DlsoBackend::~DlsoBackend()
+{
+  if (this->api->release != nullptr)
+    this->api->release(this->api);
+  dlclose(this->dlhandle);
+}
+
+/**
+ * The functions here are just remote json stubs that send and receive the method call
+ * data is mainly left alone, some defaults are assumed.
+ */
+void DlsoBackend::lookup(const QType& qtype, const DNSName& qdomain, int32_t domain_id, DNSPacket* pkt_p)
+{
+  if (in_query)
+    throw PDNSException("Attempt to lookup while one running");
+
+  string qname = qdomain.toString();
+  bool success;
+
+  if (pkt_p != nullptr) {
+    ComboAddress edns_or_resolver_ip = pkt_p->getRealRemote().getNetwork();
+    success = api->lookup(api->handle, qtype.getCode(), qname.size(), qname.c_str(), (sockaddr*)&edns_or_resolver_ip.sin4, domain_id);
+  }
+  else {
+    success = api->lookup(api->handle, qtype.getCode(), qname.size(), qname.c_str(), nullptr, domain_id);
+  }
+
+  if (!success)
+    throw PDNSException("Backend failed");
+}
+
+bool DlsoBackend::list(const DNSName& target, int domain_id, bool include_disabled)
+{
+  if (api->list == nullptr)
+    return false;
+
+  string qname = target.toString();
+  bool success;
+
+  success = api->list(api->handle, qname.size(), qname.c_str(), domain_id);
+
+  if (!success)
+    throw PDNSException("Backend failed");
+
+  return success;
+}
+
+void fill_cb(const void* ptr, const struct resource_record* record)
+{
+  DNSResourceRecord* rr = (DNSResourceRecord*)ptr;
+  rr->qtype = record->qtype;
+  string qname = string(record->qname, record->qname_len);
+  rr->qname = DNSName(qname);
+  rr->qclass = QClass::IN;
+  rr->content = string(record->content, record->content_len);
+  rr->ttl = record->ttl;
+  rr->auth = record->auth;
+  rr->scopeMask = record->scope_mask;
+  rr->domain_id = record->domain_id;
+}
+
+bool DlsoBackend::get(DNSResourceRecord& rr)
+{
+  bool success = api->get(api->handle, fill_cb, &rr);
+
+  if (!success) {
+    in_query = false;
+    return false;
+  }
+
+  return true;
+}
+
+void fill_meta_cb(const void* ptr, uint8_t value_len, const struct dns_value* values)
+{
+  std::vector<std::string>* meta = (std::vector<std::string>*)ptr;
```

Comment:
Same here.

Reply:
Ideally you would never use a C-style cast in C++, so you would likely need a `reinterpret_cast` and a `const_cast`.